include/ginkgo/core/stop/residual_norm.hpp

```diff
@@ -226,6 +226,122 @@ class RelativeResidualNorm : public ResidualNorm<ValueType> {
 };
 
 
+/**
+ * The ImplicitResidualNormReduction class is a stopping criterion which
+ * stops the iteration process when the implicit residual norm is below a
+ * certain threshold relative to either the norm of the right-hand side, i.e.
+ * when norm(residual) / norm(right_hand_side) < threshold or the initial
+ * residual, i.e. when norm(residual) / norm(initial_residual) < threshold. For
+ * better performance, the checks are run on the executor
+ * where the algorithm is executed.
+ *
+ * @note To use this stopping criterion there are some dependencies. The
+ * constructor depends on either `b` or the `initial_residual` in order to
+ * compute their norms. If this is not correctly provided, an exception
+ * ::gko::NotSupported() is thrown.
+ *
+ * @ingroup stop
+ */
+template <typename ValueType = default_precision>
+class ImplicitResidualNormReduction : public ResidualNorm<ValueType> {
+public:
+    using ComplexVector = matrix::Dense<to_complex<ValueType>>;
+    using NormVector = matrix::Dense<remove_complex<ValueType>>;
+    using Vector = matrix::Dense<ValueType>;
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Implicit residual norm reduction factor
+         */
+        remove_complex<ValueType> GKO_FACTORY_PARAMETER_SCALAR(
+            reduction_factor, static_cast<remove_complex<ValueType>>(1e-15));
+
+        /**
+         * The quantity the reduction is relative to. Choices include "rhs" and
+         * "initial_residual"
+         */
+        std::string GKO_FACTORY_PARAMETER_SCALAR(relative_to, "rhs");
```

Comment:
Why do you use a string for that? I would personally prefer an Enum because it is easier and more efficient to compare while also having compile-time checks if the value is valid (if not, it simply fails to compile). With a string, it is easy to mistype a name and get a runtime error, which is a lot harder to debug (especially since the only information you get is `NOT_SUPPORTED` where you need to read the file to understand where the error is).
If you only ever want to have `rhs` and `initial_residual` as options, a `bool` value might be better (`relative_to_initial_residual` might be a good name, which you already use internally).

Reply:
Good point. I use an `enum` now.